CesiumAsync/include/CesiumAsync/Impl/AsyncSystemSchedulers.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "CesiumAsync/Impl/cesium-async++.h"
+
+namespace CesiumAsync {
+
+class ITaskProcessor;
+
+namespace Impl {
+// Begin omitting doxgen warnings for Impl namespace
+//! @cond Doxygen_Suppress
+
+class AsyncSystemSchedulers;
+
+class ImmediatelyInMainThreadScheduler {
+public:
+  ImmediatelyInMainThreadScheduler(AsyncSystemSchedulers* pSchedulers)
+      : _pSchedulers(pSchedulers) {}
+
+  void schedule(async::task_run_handle t);
+
+private:
+  AsyncSystemSchedulers* _pSchedulers;
+};
+
+class WorkerThreadScheduler {
+public:
+  WorkerThreadScheduler(AsyncSystemSchedulers* pSchedulers)
+      : _pSchedulers(pSchedulers) {}
+
+  void schedule(async::task_run_handle t);
+
+private:
+  AsyncSystemSchedulers* _pSchedulers;
+};
+
+class ImmediatelyInWorkerThreadScheduler {
```

Comment:
Since `ImmediatelyInWorkerThreadScheduler` uses `WorkerThreadScheduler`, which in turn uses `ITaskProcessor` in its implementation, I think we can consolidate them together in one class. Maybe `ImmediatelyInWorkerThreadScheduler` will own `ITaskProcessor` directly and use the processor in its `schedule()` method. That way, we can remove the cyclic dependency between `AsyncSystemSchedulers` and  `ImmediatelyInWorkerThreadScheduler`. Also the WorkerThreadScheduler may not be needed.

With the change, `AsyncSystemScheduler` will still own `ImmediatelyInWorkerThreadsScheduler` and `ImmediatelyInMainThreadsScheduler`, but it won't own `ITaskProcessor` and `async::fifo_scheduler`

Reply:
Same reason as above. WorkerThreadScheduler was originally intended to be used as an independent scheduler, and making it easy to do so in the future (or at least not changing the code so it's harder!) is worthwhile.

Reply:
Similar to the one above, the proposed change mostly is to make `ImmediatelyInWorkerThreadsScheduler` not to depend on `AsyncSystemSchedulers`. I think we can keep `WorkerThreadScheduler` but let `ImmediatelyInWorkerThreadsScheduler` owns it directly and remove `AsyncSystemSchedulers` from `ImmediatelyInWorkerThreadsScheduler`

Reply:
ok so if `ImmediatelyInWorkerThreadScheduler` owns a `WorkerThreadScheduler` directly, then in order to schedule directly with the `WorkerThreadScheduler` we'd have to "reach into" the `ImmediatelyInWorkerThreadScheduler` to grab its embedded `WorkerThreadScheduler`. Or have two `WorkerThreadScheduler` instances maybe. Either way that doesn't sound like an improvement to me.

I guess the way I think about it is that these various schedulers would ideally just be methods on `AsyncSystemSchedulers`. There would be a `scheduleInWorkerThread`, `scheduleImmediatelyInWorkerThread`, etc. But to use them with Async++, we need separate instances that all have a method called `schedule`. So `AsyncSystemSchedulers` plus all the individual scheduler types are conceptually a single class. Maybe the scheduler types should be nested to make that more clear.

Reply:
Admittedly, it's not an improvement in the implementation of the `AsyncSystem`. But from the reusable point of view, I still think it is, since if someone needs to use `ImmediatelyInWorkerThreadScheduler` outside of `AsyncSystem` context, they won't need to bring in the `AsyncSystemSchedulers` which has the `ImmediatelyInMainThreadScheduler` that they don't need. But that is an exacerbation on my part as the component won't be used outside of the `AsyncSystem`, so I guess the dependency change is not worth it. Also the scheduler is simple enough to reason about the object lifetime, and it probably won't get more complicated in the future to be much of the concern.

Reply:
I'm still not really sold on the reusability angle. Or rather, given that this is an entirely private API, I think it would be wiser to refactor for reuse if and when we're actually going to expose it publicly or reuse it internally, rather than speculatively investing that time now. But I had a go at implementing some approximation of what you were describing, and I do agree it's an improvement. Let me know if this is close to what you had in mind.

Reply:
It’s better than what I had in mind actually. I agree on the reusable aspect. I realized that it’s used only in the AsyncSystem internally, so it’s unreasonable of me to ask to invest the time in it